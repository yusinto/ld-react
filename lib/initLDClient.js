"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.initLDClient = exports.ldClient = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _lodash = _interopRequireDefault(require("lodash.camelcase"));

var _ldclientJs = require("ldclient-js");

var _initUser = _interopRequireDefault(require("./initUser"));

var ldClient;
exports.ldClient = ldClient;

var initLDClient =
/*#__PURE__*/
function () {
  var _ref = (0, _asyncToGenerator2.default)(
  /*#__PURE__*/
  _regenerator.default.mark(function _callee(clientSideId) {
    var user,
        options,
        _args = arguments;
    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            user = _args.length > 1 && _args[1] !== undefined ? _args[1] : (0, _initUser.default)();
            options = _args.length > 2 ? _args[2] : undefined;
            exports.ldClient = ldClient = (0, _ldclientJs.initialize)(clientSideId, user, options);
            _context.next = 5;
            return new Promise(function (resolve) {
              return ldClient.on('ready', function () {
                var rawFlags = ldClient.allFlags();
                var flags = {};

                for (var rawFlag in rawFlags) {
                  var camelCasedKey = (0, _lodash.default)(rawFlag);
                  flags[camelCasedKey] = rawFlags[rawFlag];
                }

                resolve(flags);
              });
            });

          case 5:
            return _context.abrupt("return", _context.sent);

          case 6:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, this);
  }));

  return function initLDClient(_x) {
    return _ref.apply(this, arguments);
  };
}();

exports.initLDClient = initLDClient;
var _default = initLDClient;
exports.default = _default;