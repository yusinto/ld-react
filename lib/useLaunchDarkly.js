"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = useLaunchDarkly;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _lodash = _interopRequireDefault(require("lodash.camelcase"));

var _initLDClient = require("./initLDClient");

var _context = _interopRequireDefault(require("./context"));

function useLaunchDarkly() {
  var _useState = (0, _react.useState)(),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      flags = _useState2[0],
      setFlags = _useState2[1];
}