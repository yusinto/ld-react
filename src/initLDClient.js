import camelCase from 'lodash.camelcase';
import ldClientPackage from 'ldclient-js';
import initUser from './initUser';

export let ldClient;

export const initLDClient = async (clientSideId, user = initUser(), options) => {
  ldClient = ldClientPackage.initialize(clientSideId, user, options);

  return await new Promise(resolve => ldClient.on('ready', () => {
    const rawFlags = ldClient.allFlags();
    const flags = {};
    for (const rawFlag in rawFlags) {
      const camelCasedKey = camelCase(rawFlag);
      flags[camelCasedKey] = rawFlags[rawFlag];
    }
    resolve(flags);
  }));
};

export default initLDClient;