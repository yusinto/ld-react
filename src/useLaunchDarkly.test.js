jest.mock('react', () => {
  const original = jest.requireActual('react');
  return {
    ...original,
    useState: jest.fn(),
  }
});
jest.mock('./initLDClient', () => ({
  initLDClient: jest.fn(),
  ldClient: {
    on: jest.fn(),
  },
}));

import React, {useState} from 'react';
import TestRenderer from 'react-test-renderer';

import {initLDClient, ldClient} from './initLDClient';
import useLaunchDarkly from './useLaunchDarkly';

const clientSideId = 'deadbeef';
const flags = {'testFlag': true, 'anotherTestFlag': true};

describe('useLaunchDarkly hook', () => {
  const mockSetFlags = jest.fn();

  beforeEach(() => {
    initLDClient.mockImplementation(() => flags);
    useState.mockImplementation(() => [flags, mockSetFlags]);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test('initialise ld client and subscribe on mount', () => {
    const App = () => <div>My App</div>;
    const LaunchDarklyApp = () => useLaunchDarkly(App, ({clientSideId}));
    const component = TestRenderer.create(<LaunchDarklyApp />);

    // GOTCHA: we need to re-render a second time to trigger useEffect hooks
    // https://github.com/kentcdodds/react-testing-library/issues/215
    component.update(<LaunchDarklyApp />);

    expect(initLDClient.mock.calls[0][0]).toBe(clientSideId);
    ldClient.on.mockImplementation((e, cb) => {
      expect(e).toEqual('change');
      cb();
    });
    expect(component).toMatchSnapshot();
  });

  test('setFlags is called in both useEffect and subscribe', () => {
    const App = () => <div>My App</div>;
    const LaunchDarklyApp = () => useLaunchDarkly(App, ({clientSideId}));
    const component = TestRenderer.create(<LaunchDarklyApp />);
    component.update(<LaunchDarklyApp />);

    // first useEffect, then in subscribe
    mockSetFlags
      .mockImplementationOnce(remoteFlags => {
        expect(remoteFlags).toEqual(flags);
      })
      .mockImplementationOnce(remoteFlags => {
        expect(remoteFlags).toEqual({"anotherTestFlag": true, "testFlag": false});
      });

    ldClient.on.mockImplementation((e, cb) => {
      expect(e).toEqual('change');
      cb({'test-flag': {current: false, previous: true}});
      expect(mockSetFlags.mock.calls.length).toBe(2);
    });
  });
});