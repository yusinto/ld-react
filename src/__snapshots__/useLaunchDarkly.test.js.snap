// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useLaunchDarkly hook initialise ld client and subscribe on mount 1`] = `
<div>
  My App
</div>
`;
